db/snapshot-ctl.cc

```diff
@@ -209,13 +201,26 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
     return run_snapshot_list_operation([this] {
         return _db.map_reduce(adder<int64_t>(), [] (database& db) {
             return do_with(int64_t(0), [&db] (auto& local_total) {
-                return parallel_for_each(db.get_column_families(), [&local_total] (auto& cf_pair) {
-                    return cf_pair.second->get_snapshot_details().then([&local_total] (auto map) {
-                        for (auto&& snap_map: map) {
-                            local_total += snap_map.second.live;
+                return db.get_snapshot_details().then([&local_total] (auto&& map) {
+                    auto cf_map_it = map.cbegin();
+                    auto cf_map_end = map.cend();
+                    return do_until([&cf_map_it, &cf_map_end] { return cf_map_it == cf_map_end; }, [&cf_map_it, &cf_map_end, &local_total] {
+                        if (cf_map_it == cf_map_end || need_preempt()) {
+                            return make_ready_future<>();
                         }
-                        return make_ready_future<>();
-                     });
+
+                        auto details_it = cf_map_it->second.cbegin();
+                        auto details_end = cf_map_it->second.cend();
+                        return do_until([&details_it, &details_end] { return details_it == details_end; }, [&details_it, &details_end, &local_total] {
+                            if (details_it == details_end || need_preempt()) {
```

Comment:
Same question here. What's keeping details_it and details_end from being destroyed?